src/gui/editorwidgets/qgsrelationreferencewidget.h

```diff
@@ -213,6 +214,7 @@ class GUI_EXPORT QgsRelationReferenceWidget : public QWidget
     bool mShown = false;
     QgsRelation mRelation;
     bool mIsEditable = true;
+    QVariant mNullIndex;
```

Comment:
This is to define if the null index is valid? I find the type/variable name combination hard to interpret. The name looks like it's an integer, the type is flexible and I think in reality it's a `std::optional<bool>` which we cannot use (yet). What do you think about using a private enum (or at least a name change) here?

Reply:
> I think in reality it's a std::optional<bool> which we cannot use (yet)

Exactly. 

> What do you think about using a private enum (or at least a name change) here?

I think that a private enum would be overkill in this case, but i totally can change the name. What do you think about `mUpdateToNullIndex`. But maybe do you have a better idea in mind?

Reply:
I'm not even completely sure what it does. Would `mNullIndexRequiredUpdate` be more precise?

Reply:
> I'm not even completely sure what it doe

The aim is to have the same behavior than QGIS 2.18 when a user is changing the filter by clicking on comboboxes. For example:

![untitled](https://user-images.githubusercontent.com/9266424/39303377-cd5fe738-494d-11e8-9a52-ea3b9047b522.png)

- if the user selects `diameter`, then the FK is set to the first valid result (or `NULL` if null values are allowed). Moreover, the `material_value` combobox is not available anymore while a value is not selected in `diameter` (but it's not relevant in this case).
- if the user selects a value in `material_value` combobox, then the corresponding filter is applied and FK is the first valid value

Thus, the current `mNullIndex` allow to indicate if we're in the 1st case or the second case.

By hoping I'm not too unclear........

Reply:
Sorry, I'm completely on the wire. Can you help me with a small list explaining when `mNullIndex` is set to true, to false and to Null? It looks like it's only set to Null at initialization and after `modelUpdated` (where it has been evaluated and used), so the Null state of the mNullIndex is probably a guard against multiple runs of this codeblock?

Reply:
Yes, the Null state is a guard. Actually, as soon as the underlying model is updated,  the new slot `updateIndex` is called. However, I don't want to modify the current FK each time, but just when the model has been updated due to a user action on comboboxes. 

This way:
- `mNullIndex` is always Null except when the user click on comboboxes. So I know that I have to update the FK once the model is updated
- `mNullIndex` is True when the user click on a first item in one of the combobox (so on `brides`, `diameter` or `material_value` in the above example). So I know that I have to set the FK to Null or the first valid value (according to the mAllowValue configuration).
- `mNullIndex` is False when the user click on a value in one of the combobox. So I know that I have to select the first valid FK.

Retrospectively, I think that the member `mNullIndex` is really badly named and very confusing...

Reply:
Ah, I see. I think the variable name should somehow include the term `Filter` or `ChainFilter` and possibly also a descriptive comment (and a reminder to migrate to `std::optional`, once available ;) )

I didn't test it, just according to the description above, is the following interpretation of a possible workflow correct?

1. Select the bride 120, it will be set to the first valid entry (let's assume it's 9)
2. Select the material_value 'PVC' it's set to a possibly different entry (let's assume 10)
3. Unselect material_value again, it's set to NULL (even though the filter state is the same as in step 1)

Reply:
>  I think the variable name should somehow include the term Filter or ChainFilter and possibly also a descriptive comment (and a reminder to migrate to std::optional, once available ;) )

OK, I'll do that :).

> Unselect material_value again, it's set to NULL (even though the filter state is the same as in step 1)

Yes, it's the current behavior on 2.18 branch.

Reply:
And is that desired behavior?

Reply:
As the original author of this behavior, I can say that this was more a side-effect than intended behavior @pblottiere , if you have a chance to adjust that I don't think it will break anyones workflow but instead make the widget more predictable.

Reply:
> if you have a chance to adjust that I don't think it will break anyones workflow but instead make the widget more predictable.

OK, I'll do that.

Reply:
Hi @m-kuhn! 

I updated the code accordingly. This way, the `mNullIndex` guard is not necessary anymore :).

Let me know.